### ISSUE TITLE
Left bitwise shift operator (<<) shouldn't be treated the same as HERE Q in Bash .sh

### ISSUE THREAD

**Issue by GregTonoski** [ScintillaOrg/lexilla#137]
Steps to reproduce:
1) Type: 
```
echo $(( x << n ))
anything
anything
n
```
2) save a file as .sh

Actual result
All text after "<< n" until the next "n" is highlighted as if it was HERE Q and n was a delimiter.

Notepad++ v8.4.8 64-bit  
[https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13151](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13151)

**Comment by rdipardo** [comment]
To be more precise, it's when the shift count resembles a word that Lexilla sees a here quote:

~~~sh
echo $((1 << 8))
# => 256

BITS=8
echo $((1 << BITS))
# => 256
~~~

![LexBash-lsh-as-heredoc-q](https://user-images.githubusercontent.com/59004801/219709932-eae73c37-d1e0-411f-b6e5-f355ab3a1386.png)

This looks potentially like a straightforward fix, since binary operators are only valid in constructs that allow [arithmetic expansion], namely:

1. [double-parenthesis constructs] ("C-style" expressions) between `(( . . . ))`

   The lexer could just disallow `SCE_SH_HERE_DELIM` between `(( . . . ))`, or define a dedicated lexical state for "C-style" expressions.

2. [`let` expressions]

    ~~~sh
    let Z='1 << 8'; echo $Z
    # => 256
    ~~~

     No change needed since the expression must be quoted; `SCE_SH_STRING` doesn't nest, so covers all cases.

3. backticks:

    No change needed since `SCE_SH_BACKTICKS` doesn't nest, and the `expr` command doesn't evaluate bitwise operations anyway:

    ~~~sh
    Z=`expr 1 << 8`; echo $Z
    # bash: warning: here-document at line 1 delimited by end-of-file (wanted `8')
    # => 1
    ~~~

A complicating factor is that the interpreter doesn't require consistent spacing around the operands:

*Note: the effective shift count is only 56 here because the 2's complement of -8 pushes the result out of range*

~~~sh
echo $((1 <<-8))
# => 72057594037927936

echo $((1 <<- 8))
# => 72057594037927936

echo $((1 << -8))
# => 72057594037927936
~~~

Lexilla see two operators (a binary left-shift and an unary minus) in all but the last case, where the unary minus is taken for the start of a word:

![LexBash-neg-lsh-count-as-heredoc-q](https://user-images.githubusercontent.com/59004801/219709869-b1017385-bbb9-44f6-833d-d8737589d6e4.png)

One solution could be treating unary minus as numeric when a look-ahead finds a digit.

By the way, `TextLexers` hangs indefinitely given any Bash sample with an unterminated here-doc.

[double-parenthesis constructs]: https://tldp.org/LDP/abs/html/dblparens.html
[`let` expressions]: https://tldp.org/LDP/abs/html/internal.html#LETREF
[arithmetic expansion]: https://tldp.org/LDP/abs/html/arithexp.html

**Comment by nyamatongwe** [comment]
> By the way, TextLexers hangs indefinitely given any Bash sample with an unterminated here-doc.

`TestLexers` has a simpler implementation that didn't handle the final and later lines well. There'll be a small update to make it more similar to `CellBuffer` / `Document` from Scintilla.

**Comment by nyamatongwe** [comment]
Included in release 5.2.5.
